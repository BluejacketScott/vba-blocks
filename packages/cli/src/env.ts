import { join } from 'path';
import { homedir } from 'os';
import { Reporter, reporter } from './reporter';

export interface Env {
  isWindows: boolean;
  cwd: string;
  values: any;

  addins: string;
  scripts: string;
  cache: string;
  registry: string;
  packages: string;
  sources: string;
  staging: string | null;

  reporter: Reporter;
}

const cache = join(homedir(), '.vba-blocks');
const env: Env = {
  isWindows: process.platform === 'win32',
  cwd: process.cwd(),
  values: process.env,

  addins: join(__dirname, '../../addin/build'),
  scripts: join(__dirname, '../scripts'),
  cache,
  registry: join(cache, 'registry'),
  packages: join(cache, 'packages'),
  sources: join(cache, 'sources'),
  staging: null,

  reporter
};

export default env;
