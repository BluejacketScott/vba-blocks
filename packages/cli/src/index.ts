export { default as env } from './env';
export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest,
  loadManifest
} from './manifest';
export { Config, loadConfig } from './config';
export { default as resolve } from './resolve';
