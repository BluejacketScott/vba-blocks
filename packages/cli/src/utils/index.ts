export { default as checksum } from './checksum';
export { default as convertToToml } from './convert-to-toml';
export { default as copy } from './copy';
export { default as download } from './download';
export * from './git';
export { default as has } from './has';
export * from './is';
export { default as last } from './last';
export { default as optionList } from './option-list';
export { default as parallel } from './parallel';
export { default as plural } from './plural';
export { default as run, escape } from './run';
export { default as tmpFile } from './tmp-file';
export { default as unique } from './unique';
export { default as walk } from './walk';
export * from './zip';
