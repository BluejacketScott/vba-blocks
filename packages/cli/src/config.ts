import { join } from 'path';
import { parse as parseToml } from 'toml';
import env from './env';
import { pathExists, readFile } from './utils';
import {
  Sources,
  RegistrySource,
  PathSource,
  GitSource,
  UnsupportedSource
} from './sources';

export type Registry =
  | {}
  | { [name: string]: { index: string; packages: string } };

export interface Flags {
  git?: boolean;
  path?: boolean;
}

export interface Config {
  registry: Registry;
  flags: Flags;
  sources: Sources;
}

export interface ConfigValue {
  registry?: Registry;
  flags?: Flags;
}

const empty: ConfigValue = { registry: {}, flags: {} };
const defaults: ConfigValue = {
  registry: {
    'vba-blocks': {
      index: 'https://github.com/vba-blocks/registry',
      packages: 'https://packages.vba-blocks.com'
    }
  },
  flags: { git: true, path: true }
};

/**
 * Load config, from local, user, and environment values
 *
 * - Search for .vba-blocks/config.toml up from cwd
 * - Load ~/.vba-blocks/config.toml
 * - Load VBA_BLOCKS_* from environment
 */
export async function loadConfig(): Promise<Config> {
  const user: ConfigValue = {
    ...empty,
    ...((await readConfig(env.cache)) || {})
  };
  const file = await findConfig(env.cwd);
  const local: ConfigValue = {
    ...empty,
    ...file ? await readConfig(file) : {}
  };
  const override = loadConfigFromEnv();

  const registry: Registry = {
    ...defaults.registry,
    ...user.registry,
    ...local.registry,
    ...override.registry
  };
  const flags: Flags = {
    ...defaults.flags,
    ...user.flags,
    ...local.flags,
    ...override.flags
  };

  const sources: Sources = {
    registry: {},
    git: flags.git ? new GitSource() : new UnsupportedSource('git'),
    path: flags.path ? new PathSource() : new UnsupportedSource('path')
  };

  for (const [name, { index, packages }] of Object.entries(registry)) {
    sources.registry[name] = new RegistrySource({
      name,
      index,
      packages
    });
  }

  return { registry, flags, sources };
}

// Read config from dir (if present)
export async function readConfig(
  dir: string
): Promise<ConfigValue | undefined> {
  const file = join(dir, 'config.toml');
  if (!await pathExists(file)) return {};

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());

  return parsed;
}

// Find config up from and including given dir
// (looking for .vba-blocks/config.toml)
export async function findConfig(dir: string): Promise<string | undefined> {
  // TODO Search from .vba-blocks/config.toml starting at cwd
  return;
}

// Load config values from environment
export function loadConfigFromEnv(): ConfigValue {
  // TODO Load override config from VBA_BLOCKS_* env variables
  return {};
}
