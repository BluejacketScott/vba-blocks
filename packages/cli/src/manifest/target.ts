import { ok } from 'assert';
import { join } from 'path';
import { has } from '../utils';

export type TargetType = 'xlsx' | 'xlsm' | 'xlam';

export interface Target {
  name: string;
  type: TargetType;
  path: string;
}

const EXAMPLE = `Example vba-block.toml:

  [[targets]]
  type = "xlsm"
  path = "targets/xlsm"

  [[targets]]
  name = "addin"
  type = "xlam"
  path = "targets/xlam"`;

export function parseTargets(
  values: any[],
  pkgName: string,
  dir: string
): Target[] {
  ok(Array.isArray(values), `targets must be an array. ${EXAMPLE}`);

  return values.map(value => parseTarget(value, pkgName, dir));
}

export function parseTarget(value: any, pkgName: string, dir: string): Target {
  if (!has(value, 'name')) value = { name: pkgName, ...value };
  const { name, type, path: relativePath } = value;

  ok(type, `target "${name}" is missing type. ${EXAMPLE}`);
  ok(
    relativePath,
    `target "${name}" (type = "${type}") is missing path. ${EXAMPLE}`
  );

  const path = join(dir, relativePath);

  return { name, type, path };
}
