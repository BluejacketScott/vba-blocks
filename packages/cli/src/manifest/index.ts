import { join } from 'path';
import { ok } from 'assert';
import { parse as parseToml } from 'toml';
import { pathExists, readFile } from '../utils/fs';
import { Version } from './version';
import { Source, parseSrc } from './source';
import { Feature, parseFeatures } from './feature';
import { Dependency, parseDependencies } from './dependency';
import { Reference, parseReferences } from './reference';
import { Target, parseTargets } from './target';

export { Version, Source, Feature, Dependency, Reference, Target };

/**
 * @example
 * ```toml
 * [package]
 * name = "package-name"
 * version = "1.0.0-rc.1"
 * authors = ["Tim Hall <tim.hall.engr@gmail.com> (https://github.com/timhall)"]
 * 
 * [src]
 * A = "src/a.bas"
 * B = { path = "src/b.cls" }
 * C = { path = "src/c.frm", optional = true }
 * 
 * [features]
 * default = ["a"]
 * 
 * a = { src = ["C"] }
 * b = { dependencies = ["dictionary"] }
 * c = { references = ["Scripting"] }
 * 
 * [dependencies]
 * dictionary = "v1.4.1"
 * with-features = { version = "1.0.0", default-features = false, features = ["other"] }
 * from-path = { path = "packages/from-path" }
 * from-git-master = { git = "https://github.com/VBA-tools/VBA-Web.git" }
 * from-git-branch = { git = "https://github.com/VBA-tools/VBA-Web.git", branch = "beta" }
 * from-git-tag = { git = "https://github.com/VBA-tools/VBA-Web.git", tag = "v1.0.0" }
 * from-git-rev = { git = "https://github.com/VBA-tools/VBA-Web.git", rev = "a1b2c3d4" }
 * 
 * [references.Scripting]
 * version = "1.0"
 * guid = "{420B2830-E718-11CF-893D-00A0C9054228}"
 * optional = true
 * 
 * [[targets]]
 * type = "xlsm"
 * path = "targets/xlsm"
 * 
 * [[targets]]
 * name = "custom-name"
 * type = "xlam"
 * path = "targets/xlam"
 * ```
 */

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;
}

export interface Manifest extends Snapshot {
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  targets: Target[];
  features: Feature[];
  defaultFeatures: string[];
  dir: string;
}

const EXAMPLE = `Example vba-block.toml:

  [package]
  name = "my-package"
  version = "0.0.0"
  authors = ["..."]`;

export function parseManifest(value: any, dir: string): Manifest {
  ok(
    value && value.package,
    `[package] is a required field, with name, version, and authors specified. ${EXAMPLE}`
  );

  const { name, version, authors, publish = false } = value.package;

  ok(name, `[package] name is a required field. ${EXAMPLE}`);
  ok(version, `[package] version is a required field. ${EXAMPLE}`);
  ok(authors, `[package] authors is a required field. ${EXAMPLE}`);

  const src = parseSrc(value.src || {}, dir);
  const { features, defaultFeatures } = parseFeatures(value.features || {});
  const dependencies = parseDependencies(value.dependencies || {});
  const references = parseReferences(value.references || {});
  const targets = parseTargets(value.targets || [], name, dir);

  const metadata = { publish, ...value.package };

  return {
    name,
    version,
    metadata,
    src,
    features,
    defaultFeatures,
    dependencies,
    references,
    targets,
    dir
  };
}

export async function loadManifest(dir: string): Promise<Manifest> {
  const file = join(dir, 'vba-block.toml');
  if (!await pathExists(file)) {
    throw new Error(`vba-blocks.toml not found in "${dir}"`);
  }

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());
  const manifest = parseManifest(parsed, dir);

  return manifest;
}
