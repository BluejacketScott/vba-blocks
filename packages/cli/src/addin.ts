import env from './env';
import { Project } from './project';
import { Target, Source, Reference } from './manifest';
import { run, unixJoin } from './utils';

export type Application = string;
export type Addin = string;

export const extensions: { [application: string]: string[] } = {
  excel: ['xlsx', 'xlsm', 'xlam']
};
export const addins: { [application: string]: string } = {
  excel: 'vba-blocks.xlsm' // TEMP vba-blocks.xlam
};

const byExtension: { [extension: string]: string } = {};
for (const [application, values] of Object.entries(extensions)) {
  for (const extension of values) {
    byExtension[extension] = application;
  }
}

/**
 * Import graph of src and references into given target
 */
export async function importGraph(
  project: Project,
  target: Target,
  graph: { src: Source[]; references: Reference[] }
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);
  const { src, references } = graph;

  await run(application, addin, 'Build.ImportGraph', {
    file,
    src,
    references
  });
}

/**
 * Export src and references from given target
 */
export async function exportTo(
  project: Project,
  target: Target,
  staging: string
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);

  const result = await run(application, addin, 'Build.ExportTo', {
    file,
    staging
  });
  console.log('export result', result);
}

/**
 * Get application, addin, and file for given target
 */
export function getTargetInfo(
  project: Project,
  target: Target
): { application: Application; addin: Addin; file: string } {
  const application = byExtension[target.type];
  if (!application) throw new Error(`Unsupported target type "${target.type}"`);

  const addin = unixJoin(env.addins, addins[application]);
  const file = unixJoin(project.paths.build, target.filename);

  return { application, addin, file };
}
