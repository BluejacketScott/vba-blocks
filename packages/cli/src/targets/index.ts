import { unixJoin } from '../utils';
import { Target } from '../manifest';
import { Project } from '../project';

export { BuildGraph, createBuildGraph } from './build-graph';
export { default as stageBuildGraph } from './stage-build-graph';
export { default as createTarget } from './create-target';
export { default as buildTarget } from './build-target';

export function getFile(project: Project, target: Target): string {
  return unixJoin(project.paths.build, `${target.name}.${target.type}`);
}

export function getBackup(project: Project, target: Target): string {
  return unixJoin(project.paths.backup, `${target.name}.${target.type}`);
}
