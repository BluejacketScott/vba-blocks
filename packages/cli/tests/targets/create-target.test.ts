import { join } from 'path';
import { remove } from 'fs-extra';
import { unzip, walk } from '../../src/utils';
import { TargetType } from '../../src/manifest/target';
import { createTarget } from '../../src/targets';
import { loadConfig } from '../../src/config';

const FIXTURES = join(__dirname, '..', 'fixtures');
const cwd = join(FIXTURES, 'build');
const build = join(cwd, 'build');
const tmp = join(cwd, 'tmp');

afterAll(async () => Promise.all([remove(build), remove(tmp)]));

test('should create target', async () => {
  const config = await loadConfig(cwd);
  const target = {
    name: 'create-target',
    type: <TargetType>'xlsm',
    path: 'targets/xlsm'
  };

  const file = await createTarget(config, target);
  expect(file).toEqual(join(build, 'create-target.xlsm'));
  expect(await walk(build)).toMatchSnapshot();

  await unzip(file, tmp);
  expect(await walk(tmp)).toMatchSnapshot();
});
