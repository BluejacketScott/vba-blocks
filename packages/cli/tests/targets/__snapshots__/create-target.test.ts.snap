// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create target 1`] = `
Array [
  "create-target.xlsm",
]
`;

exports[`should create target 2`] = `
Array [
  "[Content_Types].xml",
  "_rels/",
  "_rels/.rels",
  "docProps/",
  "docProps/app.xml",
  "docProps/core.xml",
  "xl/",
  "xl/_rels/",
  "xl/_rels/workbook.xml.rels",
  "xl/sharedStrings.xml",
  "xl/styles.xml",
  "xl/theme/",
  "xl/theme/theme1.xml",
  "xl/workbook.xml",
  "xl/worksheets/",
  "xl/worksheets/sheet1.xml",
]
`;
